src/enn_undersample/enn_undersample.jl

```diff
@@ -0,0 +1,243 @@
+"""
+Compute a boolean filter according to a keep_condition to implement edited nearest neighbors.
+
+# Arguments
+- X: A matrix where each row is treated as an observation
+- y: A vector of labels corresponding to the observations
+- k: The number of neighbors to consider while enforcing `keep_condition`
+- keep_condition="mode:: The condition that leads to cleaning a point upon violation. Takes one of "exists", "mode", "only mode" and "all"
+    - "exists": the point has at least one neighbor from the same class
+    - "mode": the class of the point is one of the most frequent classes of the neighbors (there may be many)
+    - "only mode": the class of the point is the single most frequent class of the neighbors
+    - "all": the class of the point is the same as all the neighbors
+
+# Returns
+- A boolean filter that can be used to filter the data to remove the points violating "keep_condition"
+"""
+function compute_enn_filter(
+    X::AbstractMatrix{<:Real},
+    y::AbstractVector,
+    k::Integer,
+    keep_condition::AbstractString,
+)
+    tree = BallTree(X)
+    # Find KNN over the whole data
+    knn_map, _ = knn(tree, X, k + 1, true)
+    # Convert to matrix
+    knn_matrix = hcat(knn_map...)[2:end, :]
+    # find the labels of each neighbor
+    knn_matrix_labels = y[knn_matrix]
+    # find the points to be filtered out (a binary vector to index with)
+    modes_per_point = [modes(col) for col in eachcol(knn_matrix_labels)]
+    bool_filter = ones(Bool, length(y))
+
+    for i in 1:length(y)
+        if keep_condition == "only mode"
+            bool_filter[i] = (length(modes_per_point[i]) == 1 && y[i] in modes_per_point[i])
+        elseif keep_condition == "mode"
+            bool_filter[i] = (y[i] in modes_per_point[i])
+        elseif keep_condition == "exists"
+            bool_filter[i] = (y[i] in knn_matrix_labels[:, i])
+        else
+            neigh_labels = unique(knn_matrix_labels[:, i])
+            bool_filter[i] = (y[i] in neigh_labels && length(neigh_labels) == 1)
+        end
+    end
+
+    return BitVector(bool_filter)
+end
+
+"""
+    enn_undersample(
+        X, y; k = 5, keep_condition = "mode",
+	    min_ratios = 1.0, force_min_ratios = false,
+        rng = default_rng(), try_preserve_type=true
+    )
+
+# Description
+
+Undersample a dataset by cleaning points that violate a certain condition such as
+    belonging to a different class compared to the majority of the neighbors as proposed in [1].
```

Comment:
I think this is a bit less ambiguous:

```suggestion
Undersample a dataset by removing points that violate a certain condition such as
    belonging to a different class compared to the majority of the neighbors as proposed in [1].
```

Reply:
Agree.